torchrl/modules/models/models.py

```diff
@@ -1032,3 +1039,164 @@ def forward(
 
         input = self.mlp(input)
         return self._lstm(input, hidden0_in, hidden1_in)
+
+
+class TanhActor(nn.Module):
+    def __init__(
+        self, out_features=None, depth=None, num_cells=None, activation_class=None, rnn_hidden_dim=200
+    ):
+        super().__init__()
+        self.backbone = MLP(
+            out_features=out_features,
+            depth=depth,
+            num_cells=num_cells,
+            activation_class=activation_class,
+        )
+        self.loc_linear = nn.Linear(out_features, out_features)
+        self.scale_linear = nn.Linear(out_features, out_features)
+        self.rnn_hidden_dim = rnn_hidden_dim
+
+    def forward(self, state, belief):
+        if belief is None:
+            *batch_size,_ = state.shape
+            belief = torch.zeros(*batch_size, self.rnn_hidden_dim, device=state.device)
+        hidden = self.backbone(state, belief)
+        loc = self.loc_linear(hidden)
+        scale = self.scale_linear(hidden)
+        if self.training:
+            action = TanhNormal(loc=loc, scale=scale).rsample()
+        else:
+            action = torch.tanh(loc)
+        return action
+
+class ObsEncoder(nn.Module):
+    def __init__(self, depth=32):
+        super().__init__()
+        self.encoder = nn.Sequential(
+            nn.LazyConv2d(depth, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth, depth * 2, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth * 2, depth * 4, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(depth * 4, depth * 8, 4, stride=2),
+            nn.ReLU(),
+        )
+
+    def forward(self, observation):
+        *batch_sizes, C, H, W = observation.shape
+        observation = observation.reshape(-1, C, H, W)
+        obs_encoded = self.encoder(observation)
+        latent = obs_encoded.reshape(*batch_sizes, -1)
+        return latent
+
+
+class ObsDecoder(nn.Module):
+    def __init__(self, depth=32):
+        super().__init__()
+        self.state_to_latent = nn.Sequential(
+            nn.LazyLinear(depth * 8 * 2 * 2),
+            nn.ReLU(),
+        )
+        self.decoder = nn.Sequential(
+            nn.LazyConvTranspose2d(depth * 4, 5, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth * 4, depth * 2, 5, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth * 2, depth, 6, stride=2),
+            nn.ReLU(),
+            nn.ConvTranspose2d(depth, 3, 6, stride=2),
+        )
+        self._depth = depth
+
+    def forward(self, state, rnn_hidden):
+        latent = self.state_to_latent(torch.cat([state, rnn_hidden], dim=-1))
+        *batch_sizes, D = latent.shape
+        latent = latent.reshape(-1, D, 1, 1)
+        obs_decoded = self.decoder(latent)
+        _, C, H, W = obs_decoded.shape
+        obs_decoded = obs_decoded.reshape(*batch_sizes, C, H, W)
+        return obs_decoded
+
+class RSSMPriorRollout(nn.Module):
```

Comment:
To me this would work perfectly with tensordictmodule. 
That would permit us to preallocate the tensors of the rollout, which should be more efficient than stacking the outputs.

Reply:
You mean not having RSSMPriorRollout as an nn.Module but doing the loop over a TDModule? But then how would you integrate this with a TDSequence?